mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -0,0 +1,224 @@
+    mlir::LogicalResult matchAndRewrite(AdjointOp adjoint,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        LLVM_DEBUG(dbgs() << "Adjointing the following:\n" << adjoint << "\n");
+        Location loc = adjoint.getLoc();
+        MLIRContext *ctx = adjoint.getContext();
+        assert(adjoint.getRegion().hasOneBlock());
+
+        // First, copy the classical computations directly to the target POI and build the classical
+        // value mapping dictionary.
+        auto classicalMapping = ({
+            IRMapping out;
+            Block &b = adjoint.getRegion().front();
+            for (auto i = b.begin(); i != b.end(); i++) {
+                if (i->getName().getStringRef().find("quantum") != std::string::npos) {
+                    /* Skip quantum operations */
+                }
+                else {
+                    LLVM_DEBUG(dbgs() << "classical operation: " << *i << "\n");
+                    rewriter.insert(i->clone(out));
+                }
+            }
+            out;
+        });
+
+        // Next, compute and copy the reversed quantum computation flow. The classical dependencies
+        // such as gate parameters or qubit indices are already available in `classicalMapping`.
+        auto quantumMapping = ({
+            llvm::DenseMap<Value, Value> out;
+            auto query = [&out](Value key) -> Value {
+                LLVM_DEBUG(dbgs() << "  querying: " << key << "\n");
+                auto val = out[key];
+                LLVM_DEBUG(dbgs() << "    result: " << val << "\n");
+                return val;
+            };
+            auto update = [&out](Value key, Value val) -> void {
+                LLVM_DEBUG(dbgs() << "  updating: " << key << "\n");
+                LLVM_DEBUG(dbgs() << "    to: " << val << "\n");
+                out[key] = val;
+            };
+            Block &b = adjoint.getRegion().front();
+            auto rb = std::make_reverse_iterator(b.end());
+            auto re = std::make_reverse_iterator(b.begin());
+            for (auto i = rb; i != re; i++) {
+                LLVM_DEBUG(dbgs() << "operation: " << *i << "\n");
+                if (YieldOp yield = isInstanceOf<YieldOp>(*i)) {
+                    assert(yield.getOperands().size() == 1);
+                    update(*yield.getResults().begin(), adjoint.getQreg());
+                }
+                else if (InsertOp insert = isInstanceOf<InsertOp>(*i)) {
+                    ExtractOp extract = rewriter.create<ExtractOp>(
+                        loc, insert.getQubit().getType(), query(insert.getOutQreg()),
+                        classicalMapping.lookupOrDefault(insert.getIdx()), insert.getIdxAttrAttr());
+                    update(insert.getQubit(), extract->getResult(0));
+                    update(insert.getInQreg(), out[insert.getOutQreg()]);
+                }
+                else if (CustomOp custom = isInstanceOf<CustomOp>(*i)) {
```

Comment:
We should avoid matching against specific quantum instructions here and instead use interfaces (like the `QuantumGate` interface). If the current interfaces are insufficient for this purpose, we should see if we need to adapt them or add additional ones.

Reply:
@dime10 Can we make it a subject of future PRs? I agree in general but I would like to understand the requirements better.

Reply:
What about the clone method? I think you can call it on the interface instance and also provide an IR mapping to swap out the argument values.

Reply:
Did you end implementing it?